core/include/cubos/core/ecs/dispatcher.hpp

```diff
@@ -9,81 +9,198 @@
+        /// If the specified tag doesn't exist, it is internally created.
+        /// @param tag Tag to run before.
+        void tagSetBeforeTag(const std::string& tag);
 
-        /// Puts a stage after another stage.
-        /// @param stage Stage to put after another stage.
-        /// @param after Stage to put the stage after.
-        void putStageAfter(std::string stage, std::string referenceStage);
+        /// Adds a system, and sets it as the current system for further
+        /// settings.
+        /// @param func System to add.
+        template <typename F> void addSystem(F func);
 
-        /// Puts a stage next to another stage.
-        /// @param stage Stage to put next to another stage.
-        /// @param referenceStage Stage to put the stage next to.
-        /// @param direction Direction to put the stage in.
-        void putStage(std::string stage, std::string referenceStage, Direction direction);
+        /// Sets the tag for the current system.
+        /// @param tag The tag to run under.
+        void systemSetTag(const std::string& tag);
+
+        /// Sets the current system to run after the tag.
+        /// If the specified tag doesn't exist, it is internally created.
+        /// @param tag The tag to run after.
+        void systemSetAfterTag(const std::string& tag);
+
+        /// Sets the current system to run after a given system.
+        /// The specified system must exist.
+        /// @param func The system to run after.
+        template <typename F> void systemSetAfterSystem(F func);
+
+        /// Sets the current system to run before the tag.
+        /// If the specified tag doesn't exist, it is internally created.
+        /// @param tag The tag to run before.
+        void systemSetBeforeTag(const std::string& tag);
+
+        /// Sets the current system to run before the system.
+        /// The specified system must exist.
+        /// @param func The system to run before.
+        template <typename F> void systemSetBeforeSystem(F func);
+
+        /// Compiles a call chain. This takes all pending systems and
+        /// determines their execution order. This is required before
+        /// systems can be run.
+        void compileChain();
+
+        /// Calls all systems in order of the stages they are in.
+        /// The call chain must be compiled successfully prior, otherwise
+        /// this function won't run anything.
+        /// @param world World to call the systems in.
+        /// @param cmds Command buffer.
+        void callSystems(World& world, CommandBuffer& cmds);
+
+        /// Visits a DFSNode to create a topological order.
+        /// This is used internally during call chain compilation.
+        /// @param node The node to visit.
+        /// @param nodes Array of DFSNodes.
+        /// @return True if a cycle was detected, false if otherwise.
+        bool dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes);
```

Comment:
Make this private

Reply:
I went ahead and also made the new internal classes private as well.